groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp

```diff
@@ -0,0 +1,274 @@
+        // DEFAULT CONSTRUCTION AND 'what' METHOD
+        //
+        // Concerns:
+        //: 1 A 'bad_optional_access' object can be default constructed.
+        //: 2 Invoking a 'what' method on a 'bad_optional_access' object
+        //:   returns a "bad_optional_access" string.
+        //: 3 'what' method can be invoked on a const 'bad_optional_access'
+        //:   object
+        //
+        // Plan:
+        //: 1 For concern 1, default construct a 'bad_optional_access' object.
+        //: 2 For concern 2, invoke the 'what' method on a
+        //:   'bad_optional_access' object and check that the returned string
+        //:   is "bad_optional_access".
+        //: 3 For concern 3, in step 2, use a const qualified
+        //:   'bad_optional_access' object.
+        //
+        // Testing:
+        //   bad_optional_access();
+        //   const char *what() const;
+        // --------------------------------------------------------------------
+        if (verbose) printf("\nDEFAULT CONSTRUCTION AND 'what' METHOD"
+                            "\n======================================\n");
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+        const bsl::bad_optional_access b;
+        ASSERT(0 == strcmp("bad_optional_access", b.what()));
+#endif // #ifndef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+
+      } break;
+      case 2: {
+        // --------------------------------------------------------------------
+        // 'noexcept' SPECIFICATION
+        //
+        // Concerns:
+        //: 1 The 'noexcept' specification has been applied to
+        //:   'bad_optional_access' constructor as required by the standard.
+        //: 2 The 'noexcept' specification has been applied to
+        //:   'bad_optional_access' 'what' method as required by the standard.
+        // Plan:
+        //: 1 Apply the unary 'noexcept' operator to a 'bad_optional_access'
+        //:   constructor and, for concern 1, confirm that calculated boolean
+        //:   value matches the expected value.
+        //: 2 Apply the unary 'noexcept' operator to 'bad_optional_access'
+        //:   'what' method and, for concern 2, confirm that calculated boolean
+        //:   value matches the expected value.
+        //
+        // Testing:
+        //   CONCERN: Methods qualified 'noexcept' in standard are so.
+        // --------------------------------------------------------------------
+
+        if (verbose) printf("\n'noexcept' SPECIFICATION"
+                            "\n========================\n");
+
+        // N4835: 20.6.5 Class 'bad_optional_access' [optional.bad.access]
+        //..
+        //     namespace std {
+        //       class bad_optional_access: public std::exception {
+        //       public:
+        //         const char* what() const noexcept override;
+        //       };
+        //     } // namespace std
+        //..
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
```

Comment:
This test is also not C++17 specific (but is C++11 specific)

Reply:
Same as above applies for C++17. I'm going to assume the 'but is C++11 specific' comment is related to noexcept specifier being a C++11 feature, however, both BSLS_KEYWORD_NOEXCEPT_AVAILABLE and BSLS_KEYWORD_NOEXCEPT_AVAILABLE macros are defined in such a way that this test works in all C++ versions and there is no need to check for availability of noexcept specifier. Would you still prefer a BSLS_COMPILERFEATURES_SUPPORT_NOEXCEPT guard ?

Reply:
This one also can't have C++17 guards removed. At least one implementation (gcc) doesn't have the correct noexcept specification. Issue raised with the gcc library team. Comment added instead.

Reply:
I see what you're saying - if it compiles and runs on all platforms we're all good.